src/zap/zapheaders.cpp

```diff
@@ -190,54 +193,245 @@ void ZapImage::SaveCodeManagerEntry()
+    }
 }
 
-void ZapImage::CopyWin32VersionResource()
+void ZapImage::CopyWin32Resources()
 {
-    // Copy the version resource over so it is easy to see in the dumps where the ngened module came from
-    COUNT_T cbResourceData;
-    PVOID pResourceData = m_ModuleDecoder.GetWin32Resource(MAKEINTRESOURCE(1), RT_VERSION, &cbResourceData);
+#ifdef FEATURE_PREJIT
+    if (!IsReadyToRunCompilation())
+    {
+        // When compiling a fragile NGEN image, in order to avoid the risk of regression, only copy the RT_VERSION resource over so it 
+        // is easy to see in the dumps where the ngened module came from. For R2R, we copy all resources (new behavior).
+        COUNT_T cbResourceData;
+        PVOID pResourceData = m_ModuleDecoder.GetWin32Resource(MAKEINTRESOURCE(1), RT_VERSION, &cbResourceData);
+
+        if (!pResourceData || !cbResourceData)
+            return;
+
+        ZapBlob * pVersionData = new (GetHeap()) ZapBlobPtr(pResourceData, cbResourceData);
+
+        ZapWin32ResourceDirectory* pTypeDirectory = new (GetHeap()) ZapWin32ResourceDirectory(m_pWin32ResourceSection);
+        ZapWin32ResourceDirectory* pNameDirectory = new (GetHeap()) ZapWin32ResourceDirectory(m_pWin32ResourceSection);
+        ZapWin32ResourceDirectory* pLanguageDirectory = new (GetHeap()) ZapWin32ResourceDirectory(m_pWin32ResourceSection);
+
+        pTypeDirectory->AddEntry(RT_VERSION, false, pNameDirectory, true);
+        pNameDirectory->AddEntry(MAKEINTRESOURCE(1), false, pLanguageDirectory, true);
+        pLanguageDirectory->AddEntry(MAKEINTRESOURCE(0), false, pVersionData, false);
+
+        pTypeDirectory->PlaceNodeAndDependencies(m_pWin32ResourceSection);
+
+        m_pWin32ResourceSection->Place(pVersionData);
+
+        SetDirectoryEntry(IMAGE_DIRECTORY_ENTRY_RESOURCE, m_pWin32ResourceSection);
 
-    if (!pResourceData || !cbResourceData)
         return;
+    }
+#endif
 
-    ZapBlob * pVersionData = new (GetHeap()) ZapBlobPtr(pResourceData, cbResourceData);
+    class ResourceEnumerationCallback
+    {
+        ZapImage* m_pZapImage;
+        PEDecoder* m_pModuleDecoder;
 
-    ZapVersionResource * pVersionResource = new (GetHeap()) ZapVersionResource(pVersionData);
+        std::vector<ZapNode*> m_dataEntries;
+        std::vector<ZapBlob*> m_stringEntries;
 
-    m_pWin32ResourceSection->Place(pVersionResource);
-    m_pWin32ResourceSection->Place(pVersionData);
+        ZapWin32ResourceDirectory* m_pRootDirectory;
+        ZapWin32ResourceDirectory* m_pCurrentTypesDirectory;
+        ZapWin32ResourceDirectory* m_pCurrentNamesDirectory;
+
+        bool AddResource(LPCWSTR lpszResourceType, LPCWSTR lpszResourceName)
+        {
+            COUNT_T cbResourceData;
+            PVOID pResourceData = m_pModuleDecoder->GetWin32Resource(lpszResourceName, lpszResourceType, &cbResourceData);
+            if (!pResourceData || !cbResourceData)
+                return true; // There wasn't a neutral resource, skip to next.
```

Comment:
Are we okay with taking the size on disk hit from keeping IBC data?

Reply:
Better to keep it here, as there a simple ibcmerge command that can be used to remove it.

Reply:
I think we need to have a discussion about acceptability of keeping IBC data. @sergiy-k @MichalStrehovsky Do you have an idea of how much space keeping IBC will cost? @briansull Removing IBC data from the resource file may not actually reduce the filesize. In fact, the resource algorithms that ibcmerge uses are not guaranteed to function correctly when removing resources, it just so happens that they work if the resource was added by a prior execution of ibcmerge.

Reply:
I spoke to @MeiChin-Tsai about this and she asked me to keep the IBC data out of the crossgen'd binaries for now.

Reply:
Yes. I suggest so. Because we can give people choice in the future to keep IBC data in R2R images in the future. Keeping IBC or not should be a choice. Not to having it is the default.

Reply:
Done